Move most compliance validation to pre-run

marcparadise opened the pull request:
Because it is important that when possible, the compliance run
get associated with the converge  that preceded it (via run-id) this PR updates
compliance mode to pre-validate for most common issues before the
converge occurs.

This is a change of behavior, in that previously we would
wait until it was time to send the report after the converge before
validating, which would cause the report to not get captured for current
converge (run-id) if config errors were present.

This also adds error numbers to each of the failure conditions we
detect, in order to simplify providing the operator help with resolving
the errors.

Resolves #11106 and #11105

## Types of changes
- [x] Chore (non-breaking change that does not add functionality or fix an issue)

marcparadise commented:
Not quite ready for merge, I meant to open this as a draft pending final real-world testing.

tas50 commented on lib/chef/compliance/default_attributes.rb:
```diff
@@ -38,11 +38,12 @@ module Compliance
       "quiet" => true,
 
       # Chef Inspec Compliance profiles to be used for scan of node.
       # See README.md for details
```
we might be able to add links to the docs page that will exist since we sort of know what the path will be

marcparadise replied:
This seems to refer to this section of the original README: https://github.com/chef-cookbooks/audit#configure-node

marcparadise replied:
Which I think corresponds to this (in preview that may be dated): https://deploy-preview-2870--chef-web-docs.netlify.app/chef_compliance_phase/#compliance-phase-configuration

lamont-granquist commented:
The big concern I have here is that if a bad configuration get applied by chef-client, to itself, then that will now cause runs to fail way too early for chef-client to do anything about it.  We have this problem, of course, with other config items, but they're currently pretty minimal and obvious and necessary in a very hard sense.  Is this being driven by a solid customer desire to see this?

tas50 commented:
There's a pretty easy out for any issues with the compliance phase: remove the attribute or fix the attribute with the token or server URL in it. It's all attribute controlled so any failure can be fixed by either correcting attributes or removing attributes. Why make it a hard fail. There has been a customer driven desire to have this phase be a real phase and not a bolt on experience like it was with the audit cookbook. They view the compliance phase as being just as important (sometimes more important) than their infra converges. The same way we fail the client if we can't reach the server or fetch cookbooks we should be failing compliance hard if it's configured incorrectly. Anytime the user expressly configures something it shouldn't silently fail.

lamont-granquist commented:
Okay, yeah also the PR was updated to run at run_started which is after build_node, so the user has normal attributes and policyfiles, the only thing that hasn't run yet is cookbook attributes.  And Tim pointed out to me in zoom that the way that it consumes node attributes and not just Chef::Config state gives escape hatches for users.  That all makes me less worried about it.

I do agree compliance mode should fail hard.

If we encounter any future difficulties though we could move it to after compile time before converge starts, to allow cookbook attributes to fix it, or have it just fail hard and noisy at the end.

tas50 commented:
Oh wait these attributes could be set in a cookbook. We need to account for that.

lamont-granquist commented:
oh right just for validation.  so you need to move it to converge_start then.  that'll also cover even compile time of recipes.

it might be simpler to just have it blow up loudly at the end...  otherwise someone will try to set the variables inside a lazy block somewhere and get sad.

i'm definitely on board with the idea that compliance blows up the whole run and the engineering on audit mode where by default it was all compartmentalized was terrible.  the compliance phase is just the last phase in the run and if it throws the whole thing fails is entirely fine with me.

lamont-granquist commented:
(if we don't already by default fail hard at the end of the run talking to the data collector we probably should do that as well)

marcparadise commented:
This has been updated to move the validations to occur at `converge_start`, to ensure that all cookbook attributes that could affect validations are loaded as well. Also updated cspell.json and fixed the other found spellcheck issue.
